README.md
## NAME

    MooX::ConfigFromFile - Moo eXtension for initializing objects from
    config file

## SYNOPSIS

       package Role::Action;

       use Moo::Role;

       has operator => ( ... );

       package Action;

       use Moo;
       use MooX::ConfigFromFile; # imports the MooX::ConfigFromFile::Role

       with "Role::Action";

       sub operate {
           my $self = shift;

           return say $self->operator;
       }

       package OtherAction;

       use Moo;

       with "Role::Action", "MooX::ConfigFromFile::Role";

       sub operate {
           my $self = shift;

           return warn $self->operator;
       }

       package QuiteOtherOne;

       use Moo;

       use MooX::ConfigFromFile; # imports the MooX::ConfigFromFile::Role

       with "Role::Action";

       sub _build_config_prefix { "die" }

       sub operate {
           my $self = shift;

           return die $self->operator;
       }

       package main;

       my $action = Action->new(); # tries to find a config file in config_dirs and loads it
       my $other = OtherAction->new( config_prefix => "warn" ); # use another config file
       my $quite_o = QuiteOtherOne->new(); # quite another way to have an individual config file

## DESCRIPTION

    This module is intended to easy load initialization values for
    attributes on object construction from an appropriate config file. The
    building is done in MooX::ConfigFromFile::Role - using
    MooX::ConfigFromFile ensures the role is applied.

## AUTHOR

    Jens Rehsack, "<rehsack at cpan.org>"

## BUGS

    Please report any bugs or feature requests to "bug-moox-configfromfile
    at rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=MooX-ConfigFromFile>. I
    will be notified, and then you'll automatically be notified of progress
    on your bug as I make changes.

## SUPPORT

    You can find documentation for this module with the perldoc command.

        perldoc MooX::ConfigFromFile

    You can also look for information at:

    *   RT: CPAN's request tracker (report bugs here)

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=MooX-ConfigFromFile>

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/MooX-ConfigFromFile>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/MooX-ConfigFromFile>

    *   Search CPAN

        <http://search.cpan.org/dist/MooX-ConfigFromFile/>

## LICENSE AND COPYRIGHT

    Copyright 2013-2014 Jens Rehsack.

    This program is free software; you can redistribute it and/or modify it
    under the terms of either: the GNU General Public License as published
    by the Free Software Foundation; or the Artistic License.

    See <http://dev.perl.org/licenses/> for more information.



Changes
Revision history for MooX-ConfigFromFile

0.006	2015-03-04
    - add rough support for multiple prefixes
    - fix incorrect meta-data and incomplete handling of newer eumm features
    - improve test coverage
    - simplify some code to avoid useless conditions
    - add new tests for MooX::Cmd and MooX::Options integration
    - introduce import setting for config_identifier (consumed attribute
      from MooX::File::ConfigDir)
    - bump copyright year to 2015

0.005	2014-12-01
    - Update Makefile.PL for up-to-date resources and requirements spec
    - consolidate Synopsis in main pod
    - move from README to README.md
    - improve author tests before releasing

0.004	2014-10-28
    - add ability for options to importer of MooX::ConfigFromFile
    - add option to have a singleton config (loaded once).
    - improve documentation
    - harmonize tests

0.003	2014-08-02
    - fix links in documentation to fix RT#97429 (thanks abraxxa for
      reporting)
    - switch to ExtUtils::MakeMaker - it better maintained these days

0.002	2013-09-27
    - improve documentation
    - add test to verify we're working on Moose, too
    - ensure we're only $with->() for classes, not roles
    - add a Build.PL auto_feature for author test requires

0.001	2013-09-26
    - First version, released on an unsuspecting world.


t/testerr.pm
use strict;
use warnings FATAL => "all";

require Role::Tiny;

eval sprintf( <<'EOCDECL', ($main::OO) x 1 );
{
    package #
      Must::Fail;

    use MooX::ConfigFromFile;

    sub new { bless {}, shift }
}

{
    package #
      Already::There;

    use %s;

    sub _initialize_from_config {}

    use MooX::ConfigFromFile;
}
EOCDECL

note $main::OO;

my $mf = Must::Fail->new;
ok( !$mf->can("_initialize_from_config"), "Failed to apply MooX::ConfigFromFile::Role" );

my $ar = Already::There->new;
ok( !Role::Tiny::does_role($ar, "MooX::ConfigFromFile::Role"), "Skipped applying MooX::ConfigFromFile::Role" );


t/testlib.pm
use strict;
use warnings FATAL => "all";

use FindBin;

eval sprintf( <<'EOCDECL', ($main::OO) x 6 );
{
    package    #
      Calc::Role::BinaryOperation;
    use %s::Role;

    has a => (
        is       => "ro",
        required => 1,
    );

    has b => (
        is       => "ro",
        required => 1,
    );
}

{
    package    #
      Calc::add;
    use %s;
    use MooX::ConfigFromFile;

    with "Calc::Role::BinaryOperation";

    sub execute
    {
        my $self = shift;
        return $self->a + $self->b;
    }
}

{
    package    #
      Calc::sub;
    use %s;
    use MooX::ConfigFromFile config_prefix => "calc-operands";

    with "Calc::Role::BinaryOperation";

    sub execute
    {
        my $self = shift;
        return $self->a - $self->b;
    }
}

{
    package    #
      Calc::mul;
    use %s;
    use MooX::ConfigFromFile
      config_prefix    => "calc-operands",
      config_singleton => 1;

    with "Calc::Role::BinaryOperation";

    sub execute
    {
        my $self = shift;
        return $self->a * $self->b;
    }
}

{
    package    #
      Calc::div;
    use %s;
    use MooX::ConfigFromFile
      config_singleton => 0;

    with "Calc::Role::BinaryOperation";

    around BUILDARGS => sub {
        my $next          = shift;
        my $class         = shift;
	my $a             = shift;
	my $b             = shift;
        my $loaded_config = { a => $a, b => $b };
        $class->$next(
            loaded_config => $loaded_config,
        );
    };

    sub execute
    {
        my $self = shift;
        return $self->a / $self->b;
    }
}

{
    package    #
      Dumb::Cfg;

    use %s;
    use MooX::ConfigFromFile;

    sub execute
    {
        return;
    }
}
EOCDECL

# This is for warn once
note $main::OO;

my $adder = Calc::add->new( config_prefix => "calc-operands" );
ok( defined( $adder->a ), "read \"a\" from add config" );
ok( defined( $adder->b ), "read \"b\" from add config" );
cmp_ok( $adder->execute, "==", 5, "read right adder config" );
ok( Moo::Role::does_role( $adder, "MooX::ConfigFromFile::Role" ), "Applying MooX::ConfigFromFile::Role" );

foreach my $copy_attr ( map { "config_$_" } qw(extensions dirs prefix_map_separator prefixes prefix_map files_pattern files) )
{
    my $subber = Calc::sub->new( $copy_attr => $adder->$copy_attr );
    ok( defined( $subber->a ), "read \"a\" from sub config using $copy_attr" );
    ok( defined( $subber->b ), "read \"b\" from sub config using $copy_attr" );
    cmp_ok( $subber->execute, "==", -1, "read right subber config using $copy_attr" );
}

my $mul1 = Calc::mul->new( b => 4 );
ok( defined( $mul1->a ), "read \"a\" from mul1 config" );
ok( defined( $mul1->b ), "read \"b\" from mul1 config" );
cmp_ok( $mul1->execute, "==", 8, "read right mul config" );

my $mul2 = Calc::mul->new( config_prefix => "no-calc-operands" );
ok( defined( $mul2->a ), "copy \"a\" from mul1 config" );
ok( defined( $mul2->b ), "copy \"b\" from mul1 config" );
cmp_ok( $mul2->execute, "==", 6, "right mul2 config duplicated" );

my $mul3 = $mul2->new;
cmp_ok( $mul3->execute, "==", 6, "right mul3 config duplicated" );

my $div1 = Calc::div->new( 12, 3 );
ok( defined( $div1->a ), "read \"a\" from div1 config" );
ok( defined( $div1->b ), "read \"b\" from div1 config" );
cmp_ok( $div1->execute, "==", 4, "read right div1 config" );

my $div2 = Calc::div->new( 12, 6 );
ok( defined( $div2->a ), "read \"a\" from div2 config" );
ok( defined( $div2->b ), "read \"b\" from div2 config" );
cmp_ok( $div2->execute, "==", 2, "read right div2 config" );

my $dumb = Dumb::Cfg->new( config_dirs => 1 );
isa_ok( $dumb, "Dumb::Cfg" );
is( $dumb->config_prefix, $FindBin::Script, "fallback config prefix" );
is_deeply( $dumb->config_dirs,     [qw(.)],            "fallback config dirs" );
is_deeply( $dumb->config_prefixes, [$FindBin::Script], "fallback config prefix" );


t/testmxcmd.pm
use MooX::Cmd::Tester;

use FindBin qw($Bin);
use lib "$Bin/lib";

use MooXCmdTest;

my @tests = (
    [ [], "MooXCmdTest", "MooXCmdTest", { complicated_setting => { say => "Hello!" } } ],
    [ [qw(test)], "MooXCmdTest", "MooXCmdTest::Cmd::Test", { unintialized_attribute => sub { time - $_[0] < 1 } } ],
    [ [qw(tested)], "MooXCmdTest", "MooXCmdTest::Cmd::Tested", { confidential_setting => 42 } ],
    [ [qw(test this)], "MooXCmdTest", "MooXCmdTest::Cmd::Test::Cmd::This", { dedicated_setting => 4711 } ],
);

for (@tests)
{
  SKIP:
    {
	my ( $args, $class, $cmd_class, $attrs ) = @{$_};
	ref $args or $args = [ split( ' ', $args ) ];
	my $rv = test_cmd( $class => $args );
	#diag(explain($rv));

	my $test_ident = "$class => " . join( " ", "[", @$args, "]" );
	ok( $rv->cmd, "got cmd for $test_ident" ) or diag( explain($rv) );
	isa_ok( $rv->cmd, $class ) or skip( "Cannot do attribute testing without command", 2 );
	isa_ok( $rv->cmd->command_chain_end, $cmd_class )
	  or skip( "Cannot do attribute testing without specific command", 1 )
	  if scalar @$args;
	my $cmd = scalar @$args ? $rv->cmd->command_chain_end : $rv->cmd;
	foreach my $k ( keys %$attrs )
	{
	    my $cmd_attr = $cmd->$k;
	    ref $attrs->{$k} or is( $attrs->{$k}, $cmd_attr, "Attribute $k for $test_ident" );
	    "CODE" eq ref $attrs->{$k} and ok( $attrs->{$k}->($cmd_attr), "Attribute $k ok for $test_ident" );
	    ref $attrs->{$k}
	      and "CODE" ne ref $attrs->{$k}
	      and is_deeply( $attrs->{$k}, $cmd_attr, "Attribute $k for $test_ident" );
	}
    }
}


Build.PL
use 5.008003;
use strict;
use warnings FATAL => 'all';
use Module::Build;

my $builder = Module::Build->new(
    module_name         => 'MooX::ConfigFromFile',
    license             => 'Perl_5',
    dist_author         => q{Jens Rehsack <rehsack@cpan.org>},
    dist_version_from   => 'lib/MooX/ConfigFromFile.pm',
    release_status      => 'stable',
    configure_requires => {
        'Module::Build' => 0,
    },
    build_requires => {
	'Cwd' => 0,
	'File::Basename' => 0,
	'File::Spec' => 0,
	'JSON' => 0,
	'Test::More' => 0,
	'Test::CPAN::Changes' => 0,
	'Test::Pod::Spelling::CommonMistakes' => 0,
	'Test::CheckManifest' => 0,
    },
    requires => {
	"Config::Any" => 0,
	"File::Find::Rule" => "0.30",
	"FindBin" => 0,
	"Moo" => "1.003",
	"MooX::File::ConfigDir" => "0.002",
	"Import::Into" => 0,
	"Module::Runtime" => 0,
	"B::Hooks::EndOfScope" => 0,
    },
    auto_features => {
	author_testing => {
	    description => "Enable dependencies for author testing",
	    configure_requires => {
		'Test::CPAN::Changes' => 0,
		'Test::CheckManifest' => 0,
		'Test::Pod' => 0,
		'Test::Pod::Coverage' => 0,
		'Test::Pod::Spelling::CommonMistakes' => 0,
	    },
	},
    },
    add_to_cleanup     => [ 'MooX-ConfigFromFile-*' ],
    meta_merge =>
        { resources => { repository => "https://github.com/rehsack/MooX-ConfigFromFile", }, },
    test_files => join( ' ' => 't/*.t xt/*.t' ),

);

$builder->create_build_script();


Makefile.PL
use strict;
use warnings;

use 5.008001;

use ExtUtils::MakeMaker;

my %RUN_DEPS = (
    "Config::Any"           => 0,
    "File::Find::Rule"      => "0.30",
    "FindBin"               => 0,
    "Moo"                   => "1.003",
    "MooX::File::ConfigDir" => "0.002",
);

my %CONFIGURE_DEPS = (
    'ExtUtils::MakeMaker' => 0,
);

my %BUILD_DEPS = ();

my %TEST_DEPS = (
    'Cwd'            => 0,
    'File::Basename' => 0,
    'File::Spec'     => 0,
    'JSON'           => 0,
    'Test::More'     => 0.90,
);

WriteMakefile1(
    MIN_PERL_VERSION => '5.008001',
    META_ADD         => {
        'meta-spec' => { version => 2 },
        resources   => {
            homepage   => 'https://metacpan.org/release/MooX-ConfigFromFile',
            repository => {
                url  => 'https://github.com:rehsack/MooX-ConfigFromFile.git',
                web  => 'https://github.com/rehsack/MooX-ConfigFromFile',
                type => 'git',
            },
            bugtracker => {
                web    => 'http://rt.cpan.org/Public/Dist/Display.html?Name=MooX-ConfigFromFile',
                mailto => 'bug-MooX-ConfigFromFile@rt.cpan.org',
            },
            license => 'http://dev.perl.org/licenses/',
        },
        prereqs => {
            develop => {
                requires => {
                    'Test::CPAN::Changes'                 => 0,
                    'Test::CheckManifest'                 => 0,
                    'Module::CPANTS::Analyse'             => '0.96',
                    'Test::Kwalitee'                      => 0,
                    'Test::Pod'                           => 0,
                    'Test::Pod::Coverage'                 => 0,
                    'Test::Pod::Spelling::CommonMistakes' => 0,
                    'Test::Spelling'                      => 0,
                },
            },
            configure => {
                requires => {%CONFIGURE_DEPS},
            },
            build => { requires => {%BUILD_DEPS} },
            test  => {
                requires => {%TEST_DEPS},
                suggests => {
                    'MooX::Cmd'     => '0.012',
                    'MooX::Options' => '4.001'
                }
            },
            runtime => {
                requires => {
                    %RUN_DEPS,
                    perl => '5.8.1',
                },
            },
        },
    },
    NAME               => 'MooX::ConfigFromFile',
    VERSION_FROM       => 'lib/MooX/ConfigFromFile.pm',
    ABSTRACT_FROM      => 'lib/MooX/ConfigFromFile.pm',
    LICENSE            => 'perl',
    AUTHOR             => q{Jens Rehsack <rehsack@cpan.org>},
    PREREQ_PM          => \%RUN_DEPS,
    CONFIGURE_REQUIRES => \%CONFIGURE_DEPS,
    BUILD_REQUIRES     => \%BUILD_DEPS,
    TEST_REQUIRES      => \%TEST_DEPS,
    test               => { TESTS => 't/*.t xt/*.t' },
);

sub WriteMakefile1
{    # originally written by Alexandr Ciornii, version 0.21. Added by eumm-upgrade.
    my %params       = @_;
    my $eumm_version = $ExtUtils::MakeMaker::VERSION;
    $eumm_version = eval $eumm_version;
    die "EXTRA_META is deprecated" if ( exists( $params{EXTRA_META} ) );
    die "License not specified"    if ( !exists( $params{LICENSE} ) );
    $params{TEST_REQUIRES}
      and $eumm_version < 6.6303
      and $params{BUILD_REQUIRES} = { %{ $params{BUILD_REQUIRES} || {} }, %{ delete $params{TEST_REQUIRES} } };
    #EUMM 6.5502 has problems with BUILD_REQUIRES
    $params{BUILD_REQUIRES}
      and $eumm_version < 6.5503
      and $params{PREREQ_PM} = { %{ $params{PREREQ_PM} || {} }, %{ delete $params{BUILD_REQUIRES} } };
    ref $params{AUTHOR}
      and "ARRAY" eq ref $params{AUTHOR}
      and $eumm_version < 6.5702
      and $params{AUTHOR} = join( ", ", @{ $params{AUTHOR} } );
    delete $params{CONFIGURE_REQUIRES}    if ( $eumm_version < 6.52 );
    delete $params{MIN_PERL_VERSION}      if ( $eumm_version < 6.48 );
    delete $params{META_MERGE}            if ( $eumm_version < 6.46 );
    delete $params{META_ADD}{prereqs}     if ( $eumm_version < 6.58 );
    delete $params{META_ADD}{'meta-spec'} if ( $eumm_version < 6.58 );
    delete $params{META_ADD}              if ( $eumm_version < 6.46 );
    delete $params{LICENSE}               if ( $eumm_version < 6.31 );
    delete $params{AUTHOR}                if ( $] < 5.005 );
    delete $params{ABSTRACT_FROM}         if ( $] < 5.005 );
    delete $params{BINARY_LOCATION}       if ( $] < 5.005 );

    # more or less taken from Moose' Makefile.PL
    if ( $params{CONFLICTS} )
    {
        my $ok = CheckConflicts(%params);
        exit(0) if ( $params{PREREQ_FATAL} and not $ok );
        my $cpan_smoker = grep { $_ =~ m/(?:CR_SMOKER|CPAN_REPORTER|AUTOMATED_TESTING)/ } keys %ENV;
        unless ( $cpan_smoker || $ENV{PERL_MM_USE_DEFAULT} )
        {
            sleep 4 unless ($ok);
        }
        delete $params{CONFLICTS};
    }

    WriteMakefile(%params);
}


lib/MooX/ConfigFromFile/Role.pm
package MooX::ConfigFromFile::Role;

use strict;
use warnings;

our $VERSION = '0.007';

use Moo::Role;

use FindBin qw/$Script/;

use Config::Any;
use File::Find::Rule;

with "MooX::File::ConfigDir";

around BUILDARGS => sub {
    my $next   = shift;
    my $class  = shift;
    my $params = $class->$next(@_);
    $class->_initialize_from_config($params);
    return $params;
};

sub _initialize_from_config
{
    my ( $class, $params ) = @_;
    defined $params->{loaded_config} or $params->{loaded_config} = $class->_build_loaded_config($params);

    # This copies stuff from loaded_config into the object's parameters
    foreach my $cfg_key ( keys %{ $params->{loaded_config} } )
    {
        exists $params->{$cfg_key} and next;
        $params->{$cfg_key} = $params->{loaded_config}->{$cfg_key};
    }

    return $params;
}

has 'config_prefix' => ( is => 'lazy' );

sub _build_config_prefix { $Script; }

has 'config_prefixes' => ( is => 'lazy' );

sub _build_config_prefixes
{
    my ( $class, $params ) = @_;
    defined $params->{config_prefix} or $params->{config_prefix} = $class->_build_config_prefix($params);
    [ $params->{config_prefix} ];
}

has 'config_prefix_map_separator' => ( is => 'lazy' );

sub _build_config_prefix_map_separator { "-" }

has 'config_prefix_map' => ( is => 'lazy' );

sub _build_config_prefix_map
{
    my ( $class, $params ) = @_;

    defined $params->{config_prefix_map_separator}
      or $params->{config_prefix_map_separator} = $class->_build_config_prefix_map_separator($params);
    defined $params->{config_prefixes} or $params->{config_prefixes} = $class->_build_config_prefixes($params);

    my ( $sep, $i, @prefix_map ) = ( $params->{config_prefix_map_separator} );
    for ( $i = 0; $i < scalar @{ $params->{config_prefixes} }; ++$i )
    {
        push @prefix_map, join( $sep, @{ $params->{config_prefixes} }[ 0 .. $i ] );
    }

    \@prefix_map;
}

has 'config_extensions' => ( is => 'lazy' );

sub _build_config_extensions { [ Config::Any->extensions() ] }

has 'config_files_pattern' => ( is => 'lazy' );

sub _build_config_files_pattern
{
    my ( $class, $params ) = @_;

    defined $params->{config_prefix_map} or $params->{config_prefix_map} = $class->_build_config_prefix_map($params);
    defined $params->{config_extensions} or $params->{config_extensions} = $class->_build_config_extensions($params);
    # my @cfg_pattern = map { $params->{config_prefix} . "." . $_ } @{ $params->{config_extensions} };
    my @cfg_pattern = map {
        my $ext = $_;
        map { $_ . "." . $ext } @{ $params->{config_prefix_map} }
    } @{ $params->{config_extensions} };

    \@cfg_pattern;
}

has 'config_files' => ( is => 'lazy' );

sub _build_config_files
{
    my ( $class, $params ) = @_;

    defined $params->{config_files_pattern} or $params->{config_files_pattern} = $class->_build_config_files_pattern($params);
    defined $params->{config_dirs}          or $params->{config_dirs}          = $class->_build_config_dirs($params);
    ref $params->{config_dirs} eq "ARRAY"   or $params->{config_dirs}          = ["."];

    my @cfg_files =
      File::Find::Rule->file()->name( @{ $params->{config_files_pattern} } )->maxdepth(1)->in( @{ $params->{config_dirs} } );

    return \@cfg_files;
}

has 'loaded_config' => (
    is      => 'lazy',
    clearer => 1
);

sub _build_loaded_config
{
    my ( $class, $params ) = @_;

    defined $params->{config_files} or $params->{config_files} = $class->_build_config_files($params);
    return {} if !@{ $params->{config_files} };

    my $config = Config::Any->load_files(
        {
            files   => $params->{config_files},
            use_ext => 1
        }
    );
    my $config_merged = {};
    for my $c ( map { values %$_ } @$config )
    {
        %$config_merged = ( %$config_merged, %$c );
    }

    $config_merged;
}

=head1 NAME

MooX::ConfigFromFile::Role - Moo eXtension for initializing objects from config file

=head1 DESCRIPTION

This role adds a initializing sub around L<BUILDARGS|Moose::Manual::Construction/BUILDARGS>
and puts all what could read from config files into the hash which will be
used to construct the final object.

While it does that, it internally calls it's own _build_* methods (I<_build_config_prefix>,
I<_build_config_files> and I<_build_loaded_config>) unless the appropriate attributes are
already in C<$params>.

=head1 ATTRIBUTES

This role uses following attributes which might be suitable customized by
overloading the appropriate builder or pass defaults in construction arguments.

Be sure to read L<MooX::File::ConfigDir/ATTRIBUTES>, especially
L<MooX::File::ConfigDir/config_identifier> to understand how the L</config_dirs>
are build.

When you miss a directory - see L<File::ConfigDir/plug_dir_source> and
L<File::ConfigDir::Plack>.

=head2 config_prefix

This attribute is a string and defaults to L<FindBin>'s C<$Script>. It's
interpreted as the basename of the config file name to use.

=head2 config_prefixes

This attribute is an array of strings and defaults to C<<[ config_prefix ]>>.

=head2 config_prefix_map_separator

This attribute is a string and contains the character which is used building
I<config_prefix_map> from I<config_prefixes>.

=head2 config_prefix_map

This attribute is an array of strings containing all config-prefixes joint
together C<($0, $0.$1, $0.$1.$2, ...)> using I<config_prefix_map_separator>.

=head2 config_files_pattern

This attribute contains a cross-product of I<config_prefix_map> and
I<config_extensions>. Both are concatenated using the shell wildcard '*'.

=head2 config_dirs

This attribute is consumed from L<MooX::File::ConfigDir|MooX::File::ConfigDir/config_dirs>.
It might not be smart to override - but possible. Use with caution.

=head2 config_extensions

This attribute defaults to list of extensions from L<Config::Any|Config::Any/extensions>.

=head2 config_files

This attribute contains the list of existing files in I<config_dirs> matching
I<config_prefix> . I<config_extensions>.  Search is operated by L<File::Find::Rule>.

=head2 loaded_config

This attribute contains the config loaded while constructing the instance.
For classes set up using

  use MooX::ConfigFromFile config_singleton = 1;

this attribute is cached from the very first construction and fed by overwritten
I<builder>. The content of this attribute is passed to lower I<BUILDARGS>.

=head1 AUTHOR

Jens Rehsack, C<< <rehsack at cpan.org> >>

=head1 ACKNOWLEDGEMENTS

Toby Inkster suggested to rely on BUILDARGS instead of intercepting object
creation with nasty hacks. He also taught me a bit more how Moo(se) works.

=head1 LICENSE AND COPYRIGHT

Copyright 2013-2015 Jens Rehsack.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.

=cut

1;


lib/MooX/ConfigFromFile.pm
package MooX::ConfigFromFile;

use strict;
use warnings FATAL => 'all';

our $VERSION = '0.006';

my %loaded_configs;

sub import
{
    my ( undef, %import_options ) = @_;
    my $target = caller;
    my @target_isa;
    { no strict 'refs'; @target_isa = @{"${target}::ISA"} };

    #don't add this to a role
    #ISA of a role is always empty !
    ## no critic qw/ProhibitStringyEval/
    @target_isa or return;

    my $apply_modifiers = sub {
        return if $target->can('_initialize_from_config');
        my $with = $target->can('with');
        $with->('MooX::ConfigFromFile::Role');
    };
    $apply_modifiers->();

    my $around;
    defined $import_options{config_singleton} and $import_options{config_singleton} and do
    {
        $around = $target->can('around');
        $around->(
            _build_loaded_config => sub {
                my $orig  = shift;
                my $class = shift;
                defined $loaded_configs{$class} or $loaded_configs{$class} = $class->$orig(@_);
                return $loaded_configs{$class};
            }
        );
    };

    my %default_modifiers = (
        config_prefix               => '_build_config_prefix',
        config_identifier           => '_build_config_identifier',
        config_prefix_map_separator => '_build_config_prefix_map_separator',
        config_extensions           => '_build_config_extensions',
        config_dirs                 => '_build_config_dirs',
        config_files                => '_build_config_files',
    );

    foreach my $opt_key ( keys %default_modifiers )
    {
        exists $import_options{$opt_key} or next;
        $around or $around = $target->can('around');
        $around->( $default_modifiers{$opt_key} => sub { $import_options{$opt_key} } );
    }

    return;
}

=head1 NAME

MooX::ConfigFromFile - Moo eXtension for initializing objects from config file

=head1 SYNOPSIS

   package Role::Action;

   use Moo::Role;

   has operator => ( is => "ro" );

   package Action;

   use Moo;
   use MooX::ConfigFromFile; # imports the MooX::ConfigFromFile::Role

   with "Role::Action";

   sub operate { return say shift->operator; }

   package OtherAction;

   use Moo;

   with "Role::Action", "MooX::ConfigFromFile::Role";

   sub operate { return warn shift->operator; }

   package QuiteOtherOne;

   use Moo;

   # consumes the MooX::ConfigFromFile::Role but load config only once
   use MooX::ConfigFromFile config_singleton => 1;

   with "Role::Action";

   sub _build_config_prefix { "die" }

   sub operate { return die shift->operator; }

   package main;

   my $action = Action->new(); # tries to find a config file in config_dirs and loads it
   my $other = OtherAction->new( config_prefix => "warn" ); # use another config file
   my $quite_o = QuiteOtherOne->new(); # quite another way to have an individual config file

=head1 DESCRIPTION

This module is intended to easy load initialization values for attributes
on object construction from an appropriate config file. The building is
done in L<MooX::ConfigFromFile::Role> - using MooX::ConfigFromFile ensures
the role is applied.

For easier usage, with 0.004, several options can be passed via I<use> resulting
in default initializers for appropriate role attributes:

=over 4

=item C<config_prefix>

Default for L<MooX::ConfigFromFile::Role/config_prefix>.

=item C<config_prefix_map_separator>

Default for L<MooX::ConfigFromFile::Role/config_prefix_map_separator>.

=item C<config_extensions>

Default for L<MooX::ConfigFromFile::Role/config_extensions>.

=item C<config_dirs>

Default for L<MooX::ConfigFromFile::Role/config_dirs>.
Same warning regarding modifying this attribute applies here:
Possible, but use with caution!

=item C<config_files>

Default for L<MooX::ConfigFromFile::Role/config_files>.

Reasonable when you want exactly one config file in development mode.
For production code it is highly recommended to override the builder.

=item C<config_singleton>

Flag adding a wrapper L<< around|Class::Method::Modifiers/around method(s) => sub { ... }; >>
the I<builder> of L<MooX::ConfigFromFile::Role/loaded_config> to ensure a
config is loaded only once per class. The I<per class> restriction results
from applicable modifiers per class (and singletons are per class).

=item C<config_identifier>

Default for L<MooX::File::ConfigDir/config_identifier>.

=back

=head1 AUTHOR

Jens Rehsack, C<< <rehsack at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to
C<bug-moox-configfromfile at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=MooX-ConfigFromFile>.
I will be notified, and then you'll automatically be notified of progress
on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc MooX::ConfigFromFile

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=MooX-ConfigFromFile>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/MooX-ConfigFromFile>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/MooX-ConfigFromFile>

=item * Search CPAN

L<http://search.cpan.org/dist/MooX-ConfigFromFile/>

=back

=head1 ACKNOWLEDGEMENTS

=head1 LICENSE AND COPYRIGHT

Copyright 2013-2015 Jens Rehsack.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.

=cut

1;    # End of MooX::ConfigFromFile
